File: router-agent/go/internal/dispatch/dispatch.go
// Package dispatch builds the exact command strings accepted by the
// router-side forced-command dispatcher (router-agent/shared/router-dispatch/
// agent-dispatch.sh) and translates its stdout/stderr/exit-code convention
// into typed results or errors.
//
// Dispatcher grammar (see agent-dispatch.sh for the ground truth):
//
//	allow <ip> [timeout_min]   -> "OK allow <ip> <ntf-timeout>" on stdout, exit 0
//	block <ip>                 -> "OK block <ip>[ (not present)]" on stdout, exit 0 (idempotent)
//	list                       -> `nft -j list set` JSON on stdout, exit 0
//	status                     -> "OK table=... present" on stdout (exit 0), or
//	                              "ERR table missing" on stderr (exit 1)
//	anything else              -> "ERR ..." on stderr, exit 2 (validation error)
package dispatch

import (
	"context"
	"errors"
	"fmt"
	"strings"

	"router-agent/internal/sshclient"
)

// Kind classifies a DispatchError so the HTTP layer can map it to the right
// status code and error body without string-matching messages.
type Kind string

const (
	// KindRouterUnreachable means the SSH transport to the router could not
	// be established or was lost.
	KindRouterUnreachable Kind = "router_unreachable"
	// KindRouterTimeout means the SSH connect or exec exceeded the
	// configured timeout.
	KindRouterTimeout Kind = "router_timeout"
	// KindRouterCommandFailed means we got a response from the dispatcher
	// but it signalled an operational failure (non-zero exit outside the
	// documented "legitimate" cases) or returned output that doesn't match
	// the documented grammar at all.
	KindRouterCommandFailed Kind = "router_command_failed"
)

// Error is returned by every Dispatcher method on failure.
type Error struct {
	Kind    Kind
	Message string
}

func (e *Error) Error() string { return e.Message }

// Execer is the minimal surface Dispatcher needs from the SSH layer; it is
// satisfied by *sshclient.Client, and lets tests substitute a fake.
type Execer interface {
	Exec(ctx context.Context, cmd string) (stdout, stderr string, exitCode int, err error)
}

// Dispatcher sends validated commands to the router's forced-command
// dispatcher over an Execer (in production, a persistent *sshclient.Client)
// and interprets the results.
type Dispatcher struct {
	exec Execer
}

// New builds a Dispatcher around the given command executor.
func New(exec Execer) *Dispatcher {
	return &Dispatcher{exec: exec}
}

// AllowResult is the outcome of a successful "allow" command.
type AllowResult struct {
	IP         string
	TimeoutMin int
}

// BlockResult is the outcome of a successful "block" command.
type BlockResult struct {
	IP string
}

// StatusResult is the outcome of a successful "status" command. A false
// TablePresent reflects the dispatcher's own "ERR table missing" response,
// which is a legitimate observed router state, not a dispatch failure.
type StatusResult struct {
	TablePresent bool
}

func buildAllowCmd(ip string, timeoutMin int) string {
	return fmt.Sprintf("allow %s %d", ip, timeoutMin)
}

func buildBlockCmd(ip string) string {
	return fmt.Sprintf("block %s", ip)
}

// Allow sends "allow <ip> <timeoutMin>". Callers are expected to have
// already validated ip/timeoutMin against the same grammar the router
// enforces (see internal/api), so a validation error (exit 2) coming back
// here is treated defensively as KindRouterCommandFailed rather than
// re-derived into a 400.
func (d *Dispatcher) Allow(ctx context.Context, ip string, timeoutMin int) (*AllowResult, error) {
	stdout, stderr, exitCode, err := d.run(ctx, buildAllowCmd(ip, timeoutMin))
	if err != nil {
		return nil, err
	}
	if exitCode != 0 || !strings.HasPrefix(strings.TrimSpace(stdout), "OK allow ") {
		return nil, commandFailed("allow", stdout, stderr, exitCode)
	}
	return &AllowResult{IP: ip, TimeoutMin: timeoutMin}, nil
}

// Block sends "block <ip>". A dispatcher response of
// "OK block <ip> (not present)" is treated the same as an ordinary success —
// blocking an IP that isn't currently allowed is idempotent.
func (d *Dispatcher) Block(ctx context.Context, ip string) (*BlockResult, error) {
	stdout, stderr, exitCode, err := d.run(ctx, buildBlockCmd(ip))
	if err != nil {
		return nil, err
	}
	if exitCode != 0 || !strings.HasPrefix(strings.TrimSpace(stdout), "OK block ") {
		return nil, commandFailed("block", stdout, stderr, exitCode)
	}
	return &BlockResult{IP: ip}, nil
}

// List sends "list" and returns the raw `nft -j list set` JSON stdout for
// internal/nftjson to parse. The dispatcher always exits 0 for "list" (it
// falls back to `{"nftables":[]}` on its own if nft fails), so any non-zero
// exit here is unexpected.
func (d *Dispatcher) List(ctx context.Context) (string, error) {
	stdout, stderr, exitCode, err := d.run(ctx, "list")
	if err != nil {
		return "", err
	}
	if exitCode != 0 {
		return "", commandFailed("list", stdout, stderr, exitCode)
	}
	return stdout, nil
}

// Status sends "status". Per the dispatcher grammar, exit 0 with
// "OK table=... present" means the table exists; exit 1 with
// "ERR table missing" on stderr means it doesn't — both are well-formed,
// successful dispatch results from Status's point of view. Anything else is
// reported as KindRouterCommandFailed so the HTTP layer can distinguish
// "router told us something legitimate" from "the agent doesn't understand
// what the router just said".
func (d *Dispatcher) Status(ctx context.Context) (*StatusResult, error) {
	stdout, stderr, exitCode, err := d.run(ctx, "status")
	if err != nil {
		return nil, err
	}
	switch {
	case exitCode == 0 && strings.HasPrefix(strings.TrimSpace(stdout), "OK table="):
		return &StatusResult{TablePresent: true}, nil
	case exitCode == 1 && strings.Contains(stderr, "table missing"):
		return &StatusResult{TablePresent: false}, nil
	default:
		return nil, commandFailed("status", stdout, stderr, exitCode)
	}
}

// run executes cmd and translates transport-level sshclient errors into
// dispatch.Error, preserving Kind so the HTTP layer doesn't need to know
// about sshclient's sentinel errors at all.
func (d *Dispatcher) run(ctx context.Context, cmd string) (stdout, stderr string, exitCode int, derr error) {
	stdout, stderr, exitCode, err := d.exec.Exec(ctx, cmd)
	if err != nil {
		switch {
		case errors.Is(err, sshclient.ErrTimeout):
			return "", "", 0, &Error{Kind: KindRouterTimeout, Message: err.Error()}
		case errors.Is(err, sshclient.ErrUnreachable):
			return "", "", 0, &Error{Kind: KindRouterUnreachable, Message: err.Error()}
		default:
			return "", "", 0, &Error{Kind: KindRouterCommandFailed, Message: err.Error()}
		}
	}
	return stdout, stderr, exitCode, nil
}

func commandFailed(op, stdout, stderr string, exitCode int) *Error {
	return &Error{
		Kind: KindRouterCommandFailed,
		Message: fmt.Sprintf(
			"dispatcher %q exited %d: stdout=%q stderr=%q",
			op, exitCode, strings.TrimSpace(stdout), strings.TrimSpace(stderr),
		),
	}
}

File: router-agent/rust/src/dispatch.rs
//! Command-string builders and response interpretation for
//! `/etc/captive/agent-dispatch.sh`'s grammar:
//!
//!   allow <ip> [timeout_min]   (default 30; 0 = permanent)
//!   block <ip>
//!   list
//!   status
//!
//! stdout/stderr convention: `OK ...` on stdout with exit 0 on success;
//! `ERR ...` on stderr with exit 2 for a grammar/validation error the
//! dispatcher itself rejected, or exit 1 for an operational failure (e.g.
//! "table missing"). `list` is the one exception: on success it prints raw
//! `nft -j list set` JSON to stdout with no OK/ERR wrapper — callers parse
//! that separately with `crate::nft_json`.
//!
//! This module intentionally mirrors the sibling Go implementation's
//! `internal/dispatch` package: callers (internal/api handlers there,
//! `crate::api::handlers` here) are expected to have already validated
//! `ip`/`timeout_min` against the same grammar the router enforces, so a
//! validation error (exit 2) coming back from the dispatcher is treated
//! defensively as an ordinary `RouterCommandFailed`, exactly like any other
//! non-conforming response, rather than re-derived into a 400. `status` is
//! the one case where a non-zero exit (1, with "table missing") is a
//! legitimate, successful result rather than an error at all.

use crate::ssh_client::{ExecOutput, SshClient, SshError};

/// Classifies a `DispatchError` so the HTTP layer can map it to the right
/// status code and error body without string-matching messages. Mirrors
/// the Go sibling's `dispatch.Kind`.
// Variant names intentionally keep the `Router` prefix to match the Go
// sibling's `dispatch.Kind` constants (`KindRouterUnreachable`, etc.)
// verbatim, which is more valuable here than clippy's naming preference.
#[allow(clippy::enum_variant_names)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DispatchErrorKind {
    /// The SSH transport to the router could not be established or was lost.
    RouterUnreachable,
    /// The SSH connect or exec exceeded the configured timeout.
    RouterTimeout,
    /// We got a response from the dispatcher, but it signalled an
    /// operational failure outside the documented "legitimate" cases, or
    /// returned output that doesn't match the documented grammar at all.
    RouterCommandFailed,
}

#[derive(Debug, thiserror::Error)]
#[error("{message}")]
pub struct DispatchError {
    pub kind: DispatchErrorKind,
    pub message: String,
}

impl DispatchError {
    fn command_failed(op: &str, out: &ExecOutput) -> Self {
        DispatchError {
            kind: DispatchErrorKind::RouterCommandFailed,
            message: format!(
                "dispatcher {op:?} exited {}: stdout={:?} stderr={:?}",
                out.exit_status
                    .map(|c| c.to_string())
                    .unwrap_or_else(|| "?".to_string()),
                out.stdout.trim(),
                out.stderr.trim(),
            ),
        }
    }
}

impl From<SshError> for DispatchError {
    fn from(e: SshError) -> Self {
        let kind = match &e {
            SshError::Timeout => DispatchErrorKind::RouterTimeout,
            // A key-load/negotiation failure (Internal) means we can't even
            // start a session with the router — from the caller's point of
            // view that's indistinguishable from "unreachable".
            SshError::Unreachable(_) | SshError::Internal(_) => {
                DispatchErrorKind::RouterUnreachable
            }
        };
        DispatchError {
            kind,
            message: e.to_string(),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct AllowResult {
    pub ip: String,
    pub timeout_min: i64,
}

#[derive(Debug, Clone, PartialEq)]
pub struct BlockResult {
    pub ip: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct StatusResult {
    pub table_present: bool,
}

fn build_allow(ip: &str, timeout_min: i64) -> String {
    format!("allow {ip} {timeout_min}")
}

fn build_block(ip: &str) -> String {
    format!("block {ip}")
}

fn interpret_allow(
    ip: &str,
    timeout_min: i64,
    out: &ExecOutput,
) -> Result<AllowResult, DispatchError> {
    if out.exit_status == Some(0) && out.stdout.trim().starts_with("OK allow ") {
        Ok(AllowResult {
            ip: ip.to_string(),
            timeout_min,
        })
    } else {
        Err(DispatchError::command_failed("allow", out))
    }
}

fn interpret_block(ip: &str, out: &ExecOutput) -> Result<BlockResult, DispatchError> {
    // "OK block <ip> (not present)" also matches this prefix: blocking an
    // IP that isn't currently allowed is idempotent success.
    if out.exit_status == Some(0) && out.stdout.trim().starts_with("OK block ") {
        Ok(BlockResult { ip: ip.to_string() })
    } else {
        Err(DispatchError::command_failed("block", out))
    }
}

fn interpret_list(out: &ExecOutput) -> Result<String, DispatchError> {
    // The dispatcher always exits 0 for "list" (it falls back to
    // `{"nftables":[]}` on its own if nft fails), so any non-zero exit here
    // is unexpected.
    if out.exit_status == Some(0) {
        Ok(out.stdout.clone())
    } else {
        Err(DispatchError::command_failed("list", out))
    }
}

fn interpret_status(out: &ExecOutput) -> Result<StatusResult, DispatchError> {
    let stdout_trim = out.stdout.trim();
    if out.exit_status == Some(0) && stdout_trim.starts_with("OK table=") {
        return Ok(StatusResult {
            table_present: true,
        });
    }
    if out.exit_status == Some(1) && out.stderr.contains("table missing") {
        return Ok(StatusResult {
            table_present: false,
        });
    }
    Err(DispatchError::command_failed("status", out))
}

pub async fn allow(
    ssh: &SshClient,
    ip: &str,
    timeout_min: i64,
) -> Result<AllowResult, DispatchError> {
    let out = ssh.exec(&build_allow(ip, timeout_min)).await?;
    interpret_allow(ip, timeout_min, &out)
}

pub async fn block(ssh: &SshClient, ip: &str) -> Result<BlockResult, DispatchError> {
    let out = ssh.exec(&build_block(ip)).await?;
    interpret_block(ip, &out)
}

pub async fn list(ssh: &SshClient) -> Result<String, DispatchError> {
    let out = ssh.exec("list").await?;
    interpret_list(&out)
}

pub async fn status(ssh: &SshClient) -> Result<StatusResult, DispatchError> {
    let out = ssh.exec("status").await?;
    interpret_status(&out)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn out(stdout: &str, stderr: &str, exit_status: Option<u32>) -> ExecOutput {
        ExecOutput {
            stdout: stdout.to_string(),
            stderr: stderr.to_string(),
            exit_status,
        }
    }

    #[test]
    fn build_allow_formats_ip_and_timeout() {
        assert_eq!(build_allow("192.168.1.146", 30), "allow 192.168.1.146 30");
        assert_eq!(build_allow("192.168.1.146", 0), "allow 192.168.1.146 0");
    }

    #[test]
    fn build_block_formats_ip() {
        assert_eq!(build_block("192.168.1.146"), "block 192.168.1.146");
    }

    #[test]
    fn interpret_allow_success() {
        let o = out("OK allow 192.168.1.146 30m\n", "", Some(0));
        assert_eq!(
            interpret_allow("192.168.1.146", 30, &o).unwrap(),
            AllowResult {
                ip: "192.168.1.146".to_string(),
                timeout_min: 30
            }
        );
    }

    #[test]
    fn interpret_allow_validation_error_is_command_failed() {
        let o = out("", "ERR invalid timeout\n", Some(2));
        let err = interpret_allow("192.168.1.146", -1, &o).unwrap_err();
        assert_eq!(err.kind, DispatchErrorKind::RouterCommandFailed);
    }

    #[test]
    fn interpret_block_idempotent_not_present_is_success() {
        let o = out("OK block 192.168.1.50 (not present)\n", "", Some(0));
        assert_eq!(
            interpret_block("192.168.1.50", &o).unwrap(),
            BlockResult {
                ip: "192.168.1.50".to_string()
            }
        );
    }

    #[test]
    fn interpret_list_passes_through_raw_stdout_on_exit_0() {
        let o = out(r#"{"nftables":[]}"#, "", Some(0));
        assert_eq!(interpret_list(&o).unwrap(), r#"{"nftables":[]}"#);
    }

    #[test]
    fn interpret_list_nonzero_exit_is_command_failed() {
        let o = out("", "boom", Some(1));
        assert!(interpret_list(&o).is_err());
    }

    #[test]
    fn interpret_status_table_present() {
        let o = out("OK table=ip captive present\n", "", Some(0));
        assert_eq!(
            interpret_status(&o).unwrap(),
            StatusResult {
                table_present: true
            }
        );
    }

    #[test]
    fn interpret_status_table_missing_is_a_successful_result_not_an_error() {
        let o = out("", "ERR table missing\n", Some(1));
        assert_eq!(
            interpret_status(&o).unwrap(),
            StatusResult {
                table_present: false
            }
        );
    }

    #[test]
    fn interpret_status_malformed_output_is_command_failed() {
        let o = out("garbage\n", "", Some(0));
        let err = interpret_status(&o).unwrap_err();
        assert_eq!(err.kind, DispatchErrorKind::RouterCommandFailed);
    }

    #[test]
    fn ssh_timeout_maps_to_router_timeout_kind() {
        let derr: DispatchError = SshError::Timeout.into();
        assert_eq!(derr.kind, DispatchErrorKind::RouterTimeout);
    }

    #[test]
    fn ssh_unreachable_maps_to_router_unreachable_kind() {
        let derr: DispatchError = SshError::Unreachable("dial failed".to_string()).into();
        assert_eq!(derr.kind, DispatchErrorKind::RouterUnreachable);
    }
}
